docs: Add Carbon Efficient design document (#4686)
aws/karpenter-provider-aws: designs/carbon-aware.md

# Carbon Aware Karpenter: Optimizing Kubernetes Cluster Autoscaling for Environmental Sustainability
*Author: [@JacobValdemar](https://github.com/JacobValdemar)*

## Context & Problem

There is a growing concern about environmental sustainability within the context of Kubernetes cluster autoscaling. In multiple comments on [the proposal for moving Karpenter to CNCF](https://github.com/kubernetes/org/issues/4258), the move is backed because of opportunities within environmental sustainability.

I'm currently working on my master's thesis in Computer Engineering (M.Sc.Eng) at Aarhus University located in Denmark. The objective of the thesis is to enable Karpenter to minimize carbon emissions from Kubernetes clusters that run on cloud infrastructure (focus is AWS).

RFC: https://github.com/aws/karpenter/issues/4630

## Fundamentals of Green Software
I'll try to keep it simple. The reader should be familiar with the following.

A cluster's emissions is made of two elements: embodied emissions and operational emissions. To get the total emissions, one can add them togeather.

> - **Embodied carbon emissions**: Manufacturing emissions (CO2e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance

Reviewer: Do any cloud providers publish this data? Without it, the optimisation is hard to do.

Author: Unfortunately, cloud providers generally do not publish high-quality carbon emission data. However, as noted in the _Carbon emissions data source_ section, there are still valid methods to obtain carbon emission data. An example is Boaviztapi which you can interact with in the [data demo](https://dataviz.boavizta.org/cloudimpact) and read about their methodology in [this article](https://www.boavizta.org/en/blog/empreinte-de-la-fabrication-d-un-serveur).

> - **Embodied carbon emissions**: Manufacturing emissions (CO2e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance

Reviewer: ```suggestion
- **Embodied carbon emissions**: Manufacturing emissions (CO₂e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance
```

- **Operational carbon emissions**: Carbon emitted by electricity grid to produce electricity for the instance in the region where it is used, multiplied by PUE

There is a lot more to Green Software. If you want to learn more, I recommend you to visit [Green Software Practitioner](https://learn.greensoftware.foundation/) (a Green Software Foundation project - an affiliate of the Linux Foundation).

## Solution

### Feature Gate

> The feature is proposed to be controlled using a [feature gate](https://github.com/aws/karpenter/blob/main/website/content/en/docs/concepts/settings.md#feature-gates).

Reviewer: ```suggestion
The feature is proposed to be controlled using a [feature gate](https://karpenter.sh/docs/concepts/settings/#feature-gates).
```

| **Feature**  | **Default** |         **Config Key**          | **Stage** |    **Since**    | **Until** |
| :----------: | :---------: | :-----------------------------: | :-------: | :-------------: | :-------: |

> | Carbon Aware |    false    | featureGates.carbonAwareEnabled |   Alpha   | v0.31.0/v0.32.0 |           |

Reviewer: Not sure, but I'd expect:
```suggestiion
| CarbonAware |    false    | featureGates.carbonAwareEnabled |   Alpha   | v0.31.0/v0.32.0 |           |
```

### Carbon emissions data source
To be determined.

I don't know which one should be used yet. I am currently reviewing and comparing two options:
- [Cloud Carbon Footprint](https://github.com/cloud-carbon-footprint/ccf-coefficients). Extract data and create calculations.
- [Boavizta](https://github.com/Boavizta/boaviztapi). Extract data and port calculations (for AWS). [Data demo](https://datavizta.boavizta.org/cloudimpact).

I assume that we only want to use "static" data so we don't have to go out and make requests over the Internet to get real time data, since that would be expensive for performance.

### Changes to consolidation (karpenter-core)

Single Machine Consolidation (`singlemachineconsolidation.go`) and Multi Machine Consolidation (`multimachineconsolidation.go`) as well as `consolidation.go` is currently consolidating nodes to reduce costs. We want to change this when Carbon Aware is enabled. They should consolidate to minimize carbon emissions. 

I have identified two potential solutions.

> Recommendation: solution 1.

Reviewer: How I'd approach this:
- set a price per tonne (or kg) of CO₂e
  - maybe have a config option for this, defaulting to $0. A good starting value for taking environmental impact into account is $0.25 / kg
- adjust the price algorithm to account for the pollution cost supplement

That could be consolidation option 3.

Reviewer: Also, the same pricing choices should influence instance launch for reasons other than consolidation.

Author: @sftim That is actually a very good idea!! Have not thought of that before. That way one could "simulate" a CO₂e tax/price on the machines.

Author: @sftim I have added your idea as option 3. Interesting approach, I like it

			NewEmptyMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
			NewMultiMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
			NewSingleMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
		},
	}
}
```
</details>

Benefits:
- Current consolidation methods are unaffected
- Following the principle *Push back on requirements that introduces concepts for all users to solve problems for a few*

Disadvanteges:
- There might be copy-paste of code from the original consolidation methods to the carbon aware consolidators

#### Solution 2

Create carbon aware implementations of `filterByPrice`, `filterOutSameType`, `getCandidatePrices`, etc. that is used inside the functions when Carbon Aware is enabled. Usage of aforementioned functions might assume that it is price that they are getting, but in reality it is data about carbon emissions.

Benefits
- Less code copy-paste
- Improvements to original consolidation methods also improve the Carbon Aware feature

Disadvanteges:
- Has a risk of breaking undocumented invariants
- Adds complexity to the original consolidation methods

### Changes to Provisioning

> Currently, provisioning (roughly) filter instances based on requirements, sort instances by price, and launch the cheapest instance. We want to change this when Carbon Aware is enabled. It should sort instances by carbon emissions and launch the instance which has the lowest Global Warming Potential[^1].

Reviewer: See https://github.com/aws/karpenter/pull/4686/files#r1335705803

#### Solution 1

In karpenter-core, create a new method `types.go/OrderByCarbonEmissions` and use that in `nodeclaimtemplate.go/ToMachine` and `nodeclaimtemplate.go/ToNodeClaim` if Carbon Aware is enabled.

In karpenter, create a new method `CarbonAwareCreate` in `pkg/providers/instance/instance.go` that is used in `pkg/cloudprovider/cloudprovider.go/Create` when Carbon Aware is enabled.

> [^1]: The potential impact of greenhouse gases on global warming. Measured in terms of CO2e.

Reviewer: (nit)
```suggestion
[^1]: The potential impact of greenhouse gases on global warming. Measured in terms of CO₂e.
```

| :----------: | :---------: | :-----------------------------: | :-------: | :-------------: | :-------: |
| CarbonAware |    false    | featureGates.carbonAwareEnabled |   Alpha   | v0.31.0/v0.32.0 |           |

### Carbon emissions data source
To be determined.

I don't know which one should be used yet. I am currently reviewing and comparing two options:
- [Cloud Carbon Footprint](https://github.com/cloud-carbon-footprint/ccf-coefficients). Extract data and create calculations.
- [Boavizta](https://github.com/Boavizta/boaviztapi). Extract data and port calculations (for AWS). [Data demo](https://datavizta.boavizta.org/cloudimpact).

I assume that we only want to use "static" data so we don't have to go out and make requests over the Internet to get real time data, since that would be expensive for performance.

### Changes to consolidation (karpenter-core)

Single Machine Consolidation (`singlemachineconsolidation.go`) and Multi Machine Consolidation (`multimachineconsolidation.go`) as well as `consolidation.go` is currently consolidating nodes to reduce costs. We want to change this when Carbon Aware is enabled. They should consolidate to minimize carbon emissions. 

> I have identified two potential solutions.

Reviewer: ```suggestion
I have identified three potential solutions.
```

Author: Good catch 😁

I'm currently working on my master's thesis in Computer Engineering (M.Sc.Eng) at Aarhus University located in Denmark. The objective of the thesis is to enable Karpenter to minimize carbon emissions from Kubernetes clusters that run on cloud infrastructure (focus is AWS).

RFC: https://github.com/aws/karpenter/issues/4630

## Fundamentals of Green Software
I'll try to keep it simple. The reader should be familiar with the following.

A cluster's emissions is made of two elements: embodied emissions and operational emissions. To get the total emissions, one can add them togeather.

- **Embodied carbon emissions**: Manufacturing emissions (CO₂e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance
- **Operational carbon emissions**: Carbon emitted by electricity grid to produce electricity for the instance in the region where it is used, multiplied by PUE

There is a lot more to Green Software. If you want to learn more, I recommend you to visit [Green Software Practitioner](https://learn.greensoftware.foundation/) (a Green Software Foundation project - an affiliate of the Linux Foundation).

## Solution

> ### Feature Gate

Reviewer 2: Worth exploring an approach where you can configure instance prices separately from Karpenter. This would allow karpenter to be unaware of "carbon aware mode", and you simply provide it the updated prices. One option for prices would be a configmap.

```
priceOverrides:
  - name: m5.large
     zone: us-west-2a
     capacity-type: spot
     price: .001
  - name: m5.xlarge
     zone: us-west-2a
     capacity-type: spot
     price: .002
```

Author: Interesting idea - that would generally give users a lot of flexibility. 

However, if it is an override, it would not be possible to combine carbon emission price with actual price as [suggested by sftim](https://github.com/aws/karpenter/pull/4686#discussion_r1335703216) (although that is not a № 1 priority for the feature). Also, it would make it hard get precise operational carbon emissions as the [carbon intensity](https://learn.greensoftware.foundation/carbon-awareness#carbon-intensity) of the energy grid depends on the geographic region of the data center. I'll add it as an option tomorrow (it's 22.30 PM here)

Author: I have added it as option 4

I assume that we only want to use "static" data so we don't have to go out and make requests over the Internet to get real time data, since that would be expensive for performance.

#### Limitations
For both options above, there is a discrepancy between the available instances known to Karpenter and instances know to the carbon emissions data source. This means that as it is right now, it is not possible to get carbon emissions data for all instances types. This is mostly the case for new instance types such as m7g. Unfurtunately this seems to extend to around 300 out of 700 instance types. See full comparison in [this Gist](https://gist.github.com/JacobValdemar/e1342013c0f5c980126f6a1feb66b4a1).

I will attempt to eleminate this discrepancy, but it might not be possible. It will probably not always be possible to have an updated list of estimated carbon emissions for all instances as AWS continue to release new instance types. We should consider what to do with instance types that we do not have carbon emission estimates for.

> Approaches to handle this:

Reviewer 2: One option would be to model emissions as a price factor. e.g.,
1. If we don't know, multiply by 1
2. If known, multiply price by something > 1

You might want need to tune it such that 1 == your worst relative emissions machine.

Benefits
- Less code copy-paste
- Improvements to original consolidation methods also improve the Carbon Aware feature

Disadvanteges:
- Has a risk of breaking undocumented invariants
- Adds complexity to the original consolidation methods

#### Solution 3

Set a price per tonne (or kg) of CO₂e.

Maybe have a config option for this, defaulting to $0. A good starting value for taking environmental impact into account is $0.25 / kg. (source missing)

Adjust the price algorithm to account for the pollution cost supplement.

Additionally, if actual cost is no issue, we could weight them:
- 100% weighting to the inferred pollution cost
- 0% weighting to the billed cost from AWS

Also applies to provisioning.

### Changes to Provisioning

Currently, provisioning (roughly) filter instances based on requirements, sort instances by price, and launch the cheapest instance. We want to change this when Carbon Aware is enabled. It should sort instances by carbon emissions and launch the instance which has the lowest Global Warming Potential[^1].

> #### Solution 1

Author: I agree, that would be a good approach to succeed with our own prioritization. Standard Karpenter would still just use `lowest-price` right?

One thing I have a hard time understanding is how `pkg/cloudprovider/types.go/OrderByPrice` in `karpenter-core` is related to the ordering performed in `karpenter`. Would it even be necessary to do something with `OrderByPrice` within `karpenter-core`?

> #### Solution 1

Reviewer 2: No -- standard Karpenter would just use Prioritized, and the code would sort by price (regardless of price overrides)

> #### Solution 1

Author: @ellistarn I am unsure how spot instances would fit into all and any of the options. What do you think?

> #### Solution 1

Reviewer 2: Generational challenging. I'd suggest removing it from the scope of your MVP feature.

# Carbon Aware Karpenter: Optimizing Kubernetes Cluster Autoscaling for Environmental Sustainability
*Author: [@JacobValdemar](https://github.com/JacobValdemar)*

## Context & Problem
There is a growing concern about the environmental impact of Kubernetes clusters. Karpenter's opportunities within environmental sustainability is referenced in multiple comments that back [`karpenter-core`'s move to CNCF](https://github.com/kubernetes/org/issues/4258).

I am currently working on my master's thesis in Computer Engineering (Master of Science in Engineering) at Aarhus University located in Denmark. The objective of the thesis is to enable Karpenter to minimize carbon emissions from Kubernetes clusters that run on cloud infrastructure (scoped to AWS).

RFC: https://github.com/aws/karpenter/issues/4630

## Fundamentals of Green Software
I will try to keep it simple. The reader should be familiar with the following.

A cluster's emissions is made of two elements: embodied emissions and operational emissions. To get the total emissions, one can add them togeather.

> - **Embodied carbon emissions**: Manufacturing emissions (CO₂e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance

Reviewer 3: Can we clarify that "instance" in this sense refers to a percentage of a physical machine?

Author: @jackfrancis Good point. That is totally not obvious when reading the text. When we say instance lifetime we of course mean the lifetime of the physical machine that the instance is part of 👍

Author: a7134dd

- **Operational carbon emissions**: Carbon emitted by electricity grid to produce electricity for the instance in the region where it is used, multiplied by PUE

There is a lot more to Green Software. If you want to learn more, I recommend you to visit [Green Software Practitioner](https://learn.greensoftware.foundation/) (a Green Software Foundation project - an affiliate of the Linux Foundation).

## Solution

### Feature Gate
The feature is proposed to be controlled using a [feature gate](https://karpenter.sh/docs/concepts/settings/#feature-gates).

| **Feature** | **Default** |         **Config Key**          | **Stage** |    **Since**    | **Until** |
| :---------: | :---------: | :-----------------------------: | :-------: | :-------------: | :-------: |
| CarbonAware |    false    | featureGates.carbonAwareEnabled |   Alpha   | v0.32.0/v0.33.0 |           |

### Carbon emissions data source
Currently the best option is to create estimates based on the methodology used in [Boaviztapi](https://github.com/Boavizta/boaviztapi).

[Try out Boaviztapi on the Datavizta demo website](https://datavizta.boavizta.org/cloudimpact).

#### Licensing

> Boaviztapi is licensed under [`GNU Affero General Public License v3.0`](https://github.com/Boavizta/boaviztapi/blob/main/LICENSE). Therefore, as far as I know, we must license their data under the same license if used in the Karpenter repository.

Reviewer 3: Will there be conflicts in licensing their data according to these requirements and the CNCF guidance for licensing the karpenter-core source itself?

Author: @jackfrancis So I am in no way a lawyer, but I guess that there wont be licensing issues. We probably just have to put a different license notice in the top of files that is based on their works. Currently, I would expect the data with that license only to be in cloud provider repos (e.g. https://github.com/aws/karpenter) and not karpenter-core.

Reviewer 2: > So I am in no way a lawyer, but I guess that there wont be licensing issues

Famous last words.

Reviewer 4: We'd definitely need to check with AWS legal for this (while AWS owns it) and consider how this interacts with the CNCF guidance since this would probably live in the karpenter-core repo if it was gen-ed. I could also see this living in its own separate repo that provided configuration plugins to the pricing overrides that we are thinking about here

Reviewer 3: The particular license that Boaviztapi chooses to use suggests that it's using the "copyleft malware" approach (for Good, not Evil, of course) to ensure that its definitions of freedom (as in both beer and speech) are enforced upon downstream projects. So that's the only reason to bring this up. AWS and CNCF are probably going to combine w/ such a license in their own unique interesting ways. Non-zero chance there will be friction.

#### Limitations
There is a discrepancy between the available instances known to Karpenter and instances know to Boaviztapi. This means that as it is right now, it is not possible to get carbon emissions data for all instances types. This is mostly the case for new instance types such as m7g. Around 290 out of 700 instance types is missing data. See full comparison in [this Gist](https://gist.github.com/JacobValdemar/e1342013c0f5c980126f6a1feb66b4a1).

I will attempt to eleminate this discrepancy, but it might not be possible. It will probably not always be possible to have an updated list of estimated carbon emissions for all instances as AWS continue to release new instance types. We should consider what to do with instance types that we do not have carbon emission estimates for.

> Approaches to handle this:

Reviewer 3: Why not exclude instances that are not emission-priced when the `carbonAwareEnabled` feature gate is enabled? As a user if I were strict about this accounting I would not want the risk of polluting my reckoning with non-determinate inputs.

Author: @jackfrancis I agree. I also think an exclusion is the best option in case we can't estimate the emissions accurately. The method I thought about excluding them was to assigning them an absurd high price so they will never be picked voluntarily. However, I want to improve the dataset that we depend on, so that we can use as many instance types as possible. See this issue for reference https://github.com/Boavizta/boaviztapi/issues/232

Reviewer 3: My personal take is that if I were a customer I would prefer not to use something at all that isn't carbon-priced.

1. I would be biased to suspect that such a thing was more likely than not to be carbon-inefficient (i.e., a previous generation of tech)
2. I think that exclusion is a better forcing function for incentivizing such things to do the work of calculating cost, and/or marginalizing them from the marketplace (i.e., replacing them w/ newer, more carbon-efficient generations of tech)

-	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Controller {
+func NewController(ctx context.Context, clk clock.Clock, kubeClient client.Client, provisioner *provisioning.Provisioner,
+	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Controller {

+	if settings.FromContext(ctx).CarbonEfficientEnabled {
+		return &Controller{
+			clock:         clk,
+			kubeClient:    kubeClient,
+			cluster:       cluster,
+			provisioner:   provisioner,
+			recorder:      recorder,
+			cloudProvider: cp,
+			lastRun:       map[string]time.Time{},
+			deprovisioners: []Deprovisioner{
+				NewExpiration(clk, kubeClient, cluster, provisioner, recorder),
+				NewDrift(kubeClient, cluster, provisioner, recorder),
+				NewEmptiness(clk),
+				NewEmptyMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+				NewCarbonEfficientMultiMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+				NewCarbonEfficientSingleMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+			},
+		}

> +	}

Reviewer 5: Code gets outdated in a design doc, this bit already is outdated with the karpenter-core reference.

			NewDrift(kubeClient, cluster, provisioner, recorder),
			NewEmptiness(clk),
			NewEmptyMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
			NewMultiMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
			NewSingleMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
		},
	}
}
```
</details>

#### Provisioning
In `karpenter-core`, create a new method `types.go/OrderByCarbonEmissions` and use that in `nodeclaimtemplate.go/ToMachine` and `nodeclaimtemplate.go/ToNodeClaim` instead of `types.go/OrderByPrice` when Carbon Efficient is enabled.

In `karpenter`, create a new method `CarbonEfficientCreate` in `pkg/providers/instance/instance.go` that is used in `pkg/cloudprovider/cloudprovider.go/Create` instead of `pkg/providers/instance/instance.go/Create` when Carbon Efficient is enabled.

#### Considerations
1. 👍 Current consolidation methods are unaffected.
1. 👎 There might be copy-paste of code from the original consolidation methods to the carbon efficient consolidators.

### Option 2: Use Carbon Efficient filtering/sorting methods

#### Consolidation

> Create carbon efficient implementations of low-level functions like `filterByPrice`, `filterOutSameType`, `getCandidatePrices`, etc. that are used when Carbon Efficient is enabled. Usage of aforementioned functions might assume that it is price that they are getting, but in reality it is data about carbon emissions.

Reviewer 5: Can you go into further depth where the data consolidation will use or consider? How do cloud providers provide this carbon aware data to karpenter core?

#### Provisioning
Use same changes to provisioning as in [option 1](#option-1-use-carbon-aware-provisioning-and-concolidation-methods).

#### Considerations
1. 👍 Less code copy-paste.
1. 👍 Improvements to original consolidation methods also improve the Carbon Aware feature.
1. 👎 Has a risk of breaking undocumented invariants.
1. 👎 Adds complexity to the original consolidation methods.

### Option 3: Override instance price with carbon price (recommended)
Minimize carbon emissions by defining a price per kgCO₂e and override the instance price with the carbon price (USD/kgCO₂e). Using the `prioritized` launch strategy, carbon emissions will be minimized during provisioning. Consolidation will unknowingly consolidate to minimize carbon emissions.

> The carbon price will depend on on `region` and `instanceType` and assume constant resource utilization (e.g. always 80% utilization). The carbon price will be generated in a "hack" and included as consts (same method as used for generating initial pricing[^2]). The carbon price / emission estimates can be updated with new versions.

Reviewer 3: Can you clarify how "assume constant resource utilization" concept affects carbon price?

Author: This is one place I am constrained in my knowledge about Karpenter, so please correct me if this is incorrect: as far as I know, Karpenter can not provide information about current resource utilization (CPU/RAM) nor about expected utilization for nodes considered to be created from provisioning or consolidation.

To calculate the operational Global Warming Potential (kgCO2e) we must know how much power (W) that an instance ("server") consume. The power consumption is not static, but depends on how large the workload (%) is (e.g. see image). Power consumption is not a linear function of the workload. Therefore, CPU Power Consumption profiles (based on empirical data, aka. server testing) are used to estimate how much power a CPU consume depending on the workload. When calculating the operational carbon emissions we therefore should know what the current/expected workload is to be able to accurately estimate the operational footprint. However, if we do not know the workload (%) we must make an assumption about how large it is to complete the calculation.

I can elaborate further on this in the call today (Wednesday) if you would like, because the calculation is a bit complex to explain in a comment ;) You can read more about the calculation in the [Boavizta API (v.dev) docs here](https://dev.doc.api.boavizta.org/Explanations/components/cpu/#consumption-profile). 

See an illustration of the concept in this picture (AGPL-3.0, Boavizta API): 
![billede](https://github.com/aws/karpenter/assets/16527874/de35f745-d3c7-46a1-9d8e-89a1044e6168)

Another feature (added later) can be to add carbon price to instance price to simulate a [carbon tax](https://en.wikipedia.org/wiki/Carbon_tax). Administrators could configure a custom carbon price or use a default.

#### Considerations
1. 👍 Change is constrained to the pricing domain, so most of Karpenter's logic remains unaffected.
1. 👍👍 A simulated carbon tax could be appealing for *Beta* or *General Availability*[^3] as it combines the real price with the carbon price.
1. 👎 Adds complexity to the *price* concept. Price is not *just* price, but rather becomes an optimization function.
1. 👎 Depending on implementation, the `karpenter_cloudprovider_instance_type_price_estimate` metric *may* represent more than just price when Carbon Efficient is enabled.

### Option 4: Enable custom instance price overrides
Enable administrators to configure custom instance price overrides, e.g. in a ConfigMap. A configuration using emission factors (varying with region and instance type) masked as prices can be pre-generated. Administrators then copy-paste a Carbon Efficient `priceOverride` into their environment.

```yaml
priceOverrides:
  - instanceType: "m5.large"

>     region: "eu-west-1"

Reviewer 5: Price overrides is interested in multiple respects. For example there are reserved instances customers will reserve on azure and get a better contract for. So its actually cheaper for karpenter ot use those instances they got at a discount than the typical "retail pricing api" pricing data for the instance price. 

So this field of pricing overrides can be used and extended to other usecases.

Author: Yes, I have heard in multiple occasions that price overrides could be useful for various reasons. 

In my [Master's Thesis](https://github.com/JacobValdemar/carbon-efficient-karpenter-thesis), overriding the price of all existing instance types is actually what I did. It is feasible and maintainable with enough automation. I think it makes sense in _this_ context, but might be infeasible in other use cases.

    capacityType: OnDemand

>     price: 0.007712

Reviewer 4: I'd want to lean toward calling this `cost` now. We are starting to talk more about a "cost-optimization" function vs. a "price-optimization" function, especially if we are going to reason in terms of things that definitely aren't price

Author: I agree. The term `price` does not concisely express what the variable has/can/will become.

*Author: [@JacobValdemar](https://github.com/JacobValdemar)*

## Context & Problem
There is a growing concern about the environmental impact of Kubernetes clusters. Karpenter's opportunities within environmental sustainability is referenced in multiple comments that back [`karpenter-core`'s move to CNCF](https://github.com/kubernetes/org/issues/4258).

I am currently working on my master's thesis in Computer Engineering (Master of Science in Engineering) at Aarhus University located in Denmark. The objective of the thesis is to enable Karpenter to minimize carbon emissions from Kubernetes clusters that run on cloud infrastructure (scoped to AWS).

RFC: https://github.com/aws/karpenter/issues/4630

## Fundamentals of Green Software
I will try to keep it simple. The reader should be familiar with the following.

A cluster's emissions is made of two elements: embodied emissions and operational emissions. To get the total emissions, one can add them togeather.

- **Embodied carbon emissions**: Manufacturing emissions (CO₂e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance

> - **Operational carbon emissions**: Carbon emitted by electricity grid to produce electricity for the instance in the region where it is used, multiplied by PUE

Reviewer 6: Would this be different for regions that focus on more sustainable energy sources like hydroelectric or geothermal energy? Or are you forgoing the energy sources, and talking purely about the byproduct of the heat and other leftover elements in electricity grids?

An instance is a share of a physical machine, so when we refer to "instance lifetime" we mean the lifetime of the physical machine that the instance partially constitutes.

There is a lot more to Green Software. If you want to learn more, I recommend you to visit [Green Software Practitioner](https://learn.greensoftware.foundation/) (a Green Software Foundation project - an affiliate of the Linux Foundation).

## Solution

### Feature Gate
The feature is proposed to be controlled using a [feature gate](https://karpenter.sh/docs/concepts/settings/#feature-gates).

| **Feature** | **Default** |         **Config Key**          | **Stage** |    **Since**    | **Until** |
| :---------: | :---------: | :-----------------------------: | :-------: | :-------------: | :-------: |

> | CarbonAware |    false    | featureGates.carbonAwareEnabled |   Alpha   | v0.32.0/v0.33.0 |           |

Reviewer 4: Small detail: we're planning to drop the config key so this feature flag would probably have the same format a-la Kubernetes upstream. If we take the last solution, this feature flag might look like `PricingOverridesEnabled=true`

### Carbon emissions data source
Currently the best option is to create estimates based on the methodology used in [Boaviztapi](https://github.com/Boavizta/boaviztapi).

[Try out Boaviztapi on the Datavizta demo website](https://datavizta.boavizta.org/cloudimpact).

#### Licensing
Boaviztapi is licensed under [`GNU Affero General Public License v3.0`](https://github.com/Boavizta/boaviztapi/blob/main/LICENSE). Therefore, as far as I know, we must license their data under the same license if used in the Karpenter repository.

#### Limitations

> There is a discrepancy between the available instances known to Karpenter and instances know to Boaviztapi. This means that as it is right now, it is not possible to get carbon emissions data for all instances types. This is mostly the case for new instance types such as m7g. Around 290 out of 700 instance types is missing data. See full comparison in [this Gist](https://gist.github.com/JacobValdemar/e1342013c0f5c980126f6a1feb66b4a1).

Reviewer 7: ```suggestion
There is a discrepancy between the available instances known to Karpenter and instances know to Boaviztapi. This means that as it is right now, it is not possible to get carbon emissions data for all instances types. This is mostly the case for new instance types such as m7g. Around 290 out of 700 instance types are missing data. See full comparison in [this Gist](https://gist.github.com/JacobValdemar/e1342013c0f5c980126f6a1feb66b4a1).
```

> I will attempt to eleminate this discrepancy, but it might not be possible. It will probably not always be possible to have an updated list of estimated carbon emissions for all instances as AWS continue to release new instance types. We should consider what to do with instance types that we do not have carbon emission estimates for.

Reviewer 7: ```suggestion
I will attempt to eleminate this discrepancy, but it might not be possible. It will probably not always be possible to have an updated list of estimated carbon emissions for all instances as AWS continues to release new instance types. We should consider what to do with instance types that we do not have carbon emission estimates for.
```
Ignore this comment if this is for a British audience.

Approaches to handle this:
1. Estimate extremely high emissions to effectively filter out unknown instance types (recommended)

> 2. Estimate zero emissions

Reviewer 4: Zero is probably wrong. There may be a third option here which would be to do some estimation function. What on? I haven't thought through it that hard; however, I agree with @jackfrancis that the easy way out is to just exclude for now.

### Launch strategy

> To enable emission based priotization, the launch strategy should be changed from `lowest-price` to `prioritized`.

Reviewer 4: We probably want to do `capacity-optimized-prioritized` to properly handle spot

### Changes to consolidation (karpenter-core)
Single Machine Consolidation (`singlemachineconsolidation.go`) and Multi Machine Consolidation (`multimachineconsolidation.go`) as well as `consolidation.go` is currently consolidating nodes to reduce costs. We want to change this when Carbon Aware is enabled. They should consolidate to minimize carbon emissions. 

### Changes to Provisioning

> Currently, provisioning (roughly) filter instances based on requirements, sort instances by price, and launch the cheapest instance. We want to change this when Carbon Aware is enabled. It should sort instances by carbon emissions and launch the instance which has the lowest Global Warming Potential[^1].

Reviewer 2: Might we want a balanced approach to this?

Author: I think it could be interesting, like the carbon tax discussed in option 3, but I think that I something we should consider it something we add later

> ### Option 1: Use Carbon Aware provisioning and concolidation methods

Reviewer: ```suggestion
### Option 1: Use Carbon Aware provisioning and consolidation methods
```

			NewExpiration(clk, kubeClient, cluster, provisioner, recorder),
			NewDrift(kubeClient, cluster, provisioner, recorder),
			NewEmptiness(clk),
			NewEmptyMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
			NewMultiMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
			NewSingleMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
		},
	}
}
```
</details>

#### Provisioning
In `karpenter-core`, create a new method `types.go/OrderByCarbonEmissions` and use that in `nodeclaimtemplate.go/ToMachine` and `nodeclaimtemplate.go/ToNodeClaim` instead of `types.go/OrderByPrice` when Carbon Aware is enabled.

In `karpenter`, create a new method `CarbonAwareCreate` in `pkg/providers/instance/instance.go` that is used in `pkg/cloudprovider/cloudprovider.go/Create` instead of `pkg/providers/instance/instance.go/Create` when Carbon Aware is enabled.

#### Considerations
1. 👍 Current consolidation methods are unaffected.
1. 👎 There might be copy-paste of code from the original consolidation methods to the carbon aware consolidators.

### Option 2: Use Carbon Aware filtering/sorting methods

#### Consolidation

> Create carbon aware implementations of low-level functions like `filterByPrice`, `filterOutSameType`, `getCandidatePrices`, etc. that is used when Carbon Aware is enabled. Usage of aforementioned functions might assume that it is price that they are getting, but in reality it is data about carbon emissions.

Reviewer 7: ```suggestion
Create carbon aware implementations of low-level functions like `filterByPrice`, `filterOutSameType`, `getCandidatePrices`, etc. that are used when Carbon Aware is enabled. Usage of aforementioned functions might assume that it is price that they are getting, but in reality it is data about carbon emissions.
```

#### Provisioning
Use same changes to provisioning as in [option 1](#option-1-use-carbon-aware-provisioning-and-concolidation-methods).

#### Considerations
1. 👍 Less code copy-paste.
1. 👍 Improvements to original consolidation methods also improve the Carbon Aware feature.
1. 👎 Has a risk of breaking undocumented invariants.
1. 👎 Adds complexity to the original consolidation methods.

### Option 3: Override instance price with carbon price (recommended)
Minimize carbon emissions by defining a price per kgCO₂e and override the instance price with the carbon price (USD/kgCO₂e). Using the `prioritized` launch strategy, carbon emissions will be minimized during provisioning. Consolidation will unknowingly consolidate to minimize carbon emissions.

> The carbon price will depend on on `region` and `instanceType` and assume constant resource utilization (e.g. always 80% utilization). The carbon price will be generated in a "hack" and included as consts (same method as used for generating initial pricing[^2]). The carbon price / emission estimates can be updated with new versions.

Reviewer 7: ```suggestion
The carbon price will depend on `region` and `instanceType` and assume constant resource utilization (e.g. always 80% utilization). The carbon price will be generated in a "hack" and included as consts (same method as used for generating initial pricing[^2]). The carbon price / emission estimates can be updated with new versions.
```

Another feature (added later) can be to add carbon price to instance price to simulate a [carbon tax](https://en.wikipedia.org/wiki/Carbon_tax). Administrators could configure a custom carbon price or use a default.

#### Considerations
1. 👍 Change is constrained to the pricing domain, so most of Karpenter's logic remains unaffected.
1. 👍👍 A simulated carbon tax could be appealing for *Beta* or *General Availability*[^3] as it combines the real price with the carbon price.
1. 👎 Adds complexity to the *price* concept. Price is not *just* price, but rather becomes an optimization function.
1. 👎 Depending on implementation, the `karpenter_cloudprovider_instance_type_price_estimate` metric *may* represent more than just price when Carbon Aware is enabled.

### Option 4: Enable custom instance price overrides
Enable administrators to configure custom instance price overrides, e.g. in a ConfigMap. A configuration using emission factors (varying with region and instance type) masked as prices can be pre-generated. Administrators then copy-paste a Carbon Aware `priceOverride` into their environment.

```yaml
priceOverrides:
  - instanceType: "m5.large"

>     region: "eu-west-1"

Reviewer 2: I think there's some misunderstanding of region/zone going on.

    capacityType: OnDemand
    price: 0.007712
  - instanceType: "m5.xlarge"
    region: "eu-west-1"
    capacityType: OnDemand
    price: 0.015424
```

<details>
<summary>Alternative interface</summary>

> Alternatively, a more flexible interface could be:

Reviewer 2: Oddly, I think of this as more rigid.

```yaml
priceModification:
  operator: Add # Add or Override
  modifications:
    - instanceType: "m5.large"
      region: "eu-west-1"
      capacityType: OnDemand
      price: 0.007712
    - instanceType: "m5.xlarge"
      region: "eu-west-1"
      capacityType: OnDemand
      price: 0.015424
```
</details>

> A ConfigMap with price overrides for all combinations of instance types and regions will be very huge. 632 instances * 29 regions = 18,328 pairs. Four lines per pair gives a file with 73,312 lines. The file/configmap will approximately have a size of 2 MB. That exceeds the [`1 MiB` limit on ConfigMap size in Kubernetes](https://kubernetes.io/docs/concepts/configuration/configmap/#motivation).

Reviewer 2: EC2 launches new instance types regularly, and we should expect this to grow.
